rfcs/0000-changes-lifecyle.md

```diff
@@ -0,0 +1,205 @@
+
+## PR changes
+
+Every PR will have a section named "Release Notes", which will contains a user friendly explanation of what the PR contains.
+
+The compiler release notes **WILL** be the amalgamation of these section "as is".
+
+## Feature flags
+
+Any complex change that needed to be gated will need the following steps.
+
+1 - A specific github issue labeled with `feature-*` and `track-feature`. This issue should be the umbrella for everything related with this feature;
+2 - A preliminary PR enabling the feature flag should be created and linked in to the umbrella issue; This PR will enable the feature flag `--experimental <comma-separated-list>` on all tools.
+3 - As many PRs will be created and merged an normal;
+4 - When the feature is ready, a closing PR will be created and wait until the feature flag is enabled by default.
+5 - On a later date, the feature flag can be removed making the feature the default behavior of the compiler. 
+
+# Enabling features on `sway`
+
+Features can be enabled inside the `Forc.toml` file:
+
+```toml
+experimental = ["...", "..."]
+```
+
+or using the CLI
+
+```
+> forc ... --experimental some_feature,another_feature
+```
+
+These flags also need to be enabled programmatically by any compiler driver, like tests.
+
+Unlike `Rust` we will not support features inside sway code like the example below, because some features will span across multiple tools. That would demand `forc` to parse, or ask the `sway` compiler if a feature is enabled or not. 
+
+```sway
+#![enable(some_experimental_feature)]
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC intends to propose a scheme to track changes in the compiler. These changes will be categorized into two ways:
+
+1 - Non-breaking changes;
+2 - Breaking changes.
+
+Breaking changes will be any changes that, after updating the compiler version, causes:
+
+1 - Code without deprecated warning to be flagged as error;
+2 - ABI json properties to be removed or have their type changed;
+3 - SDK (rust, typescript and others) to break;
+4 - Receipt parsers to break;
+
+The following changes are **not** considered to be breaking changes:
+
+1 - Code flagged with deprecated warning to be flagged as error;
+2 - Code that required an experimental flag, but now are the default behavior to compile;
+3 - Contract ID to change;
+4 - new ABI json properties;
+
+## Minor Version Updates
+
+When a developer updates to a different minor version, it is expected no breaking changes, except for severe bugs.
+
+Minor changes will consider these to be breaking and non-breaking changes:
+
+### Breaking Changes 
+
+1 - If any new warning, error or other diagnostic are emitted;
+2 - If any ABI json properties are removed or have their type changed;
+3 - If any code using SDK (rust, typescript and others) stop working;
+4 - If any receipt cannot be parsed anymore;
+5 - If any "contract id" changes;
```

Comment:
This we should discuss. I don't think a contract compiling to a new contract id should necessarily be considered a breaking change. There is an argument for it but I don't see it as breaking because the code still works and compiles as intended.

If you want reproducible builds, you should be using the same exact version of the compiler anyways. So I don't see too much of a benefit in making this breaking.

Reply:
I agree with this, but am curious what the SwayEx devs would say. It would break some of our tests, which means it possibly breaks other sway dev's integration tests.

Reply:
Moving this to the open questions section.